# XB_0082 - Exercises Week 3 (2022-2023)

### Welcome to Introduction to Python Programming!

This Jupyter notebook aims to provide exercises for practicing your programming skills.  


# Introduction to this notebook

* This is a **personal** notebook. 
 
<div class="alert alert-danger" role="danger">
<h3>Integrity</h3>
<ul>
    <li>All the exercises and the graded assignments are to be done <b>individually</b>.</li>
    <li>You must not copy from the Internet, your friends, books, etc. If you represent other people's work as your own, then that constitutes fraud and will be reported to the Examination Committee.</li>
    <li>Making your work available to others (complicity) also constitutes fraud.</li>
</ul>
</div>

You are expected to work with Python 3 code in this notebook.

The locations where you should write your solutions can be recognized by **marker lines**, which look like this:

>`#//`
>    `BEGIN_TODO [Label]` `Description`
>
>`#//`
>    `END_TODO [Label]`

<div class="alert alert-warning" role="alert">Do NOT modify or delete these marker lines.  Keep them as they are.<br/>
<br/>
NEVER write code that is needed for grading <i>outside</i> the marked blocks.
</div>

Proceed in this notebook as follows:
* **Personalize** the notebook (see below)
* **Read** the text.
* **Fill in** your solutions between `BEGIN_TODO` and `END_TODO` marker lines.
* **Run** _all_ code cells (also the ones _without_ your code), _in linear order_ from the first code cell.


**Personalize your notebook**:
1. Copy the following three lines of code:

  ```python
  AUTHOR_NAME = 'Your Full Name'
  AUTHOR_ID_NR = '1234567'
  AUTHOR_DATE = 'YYYY-MM-DD'  # when first modified, e.g. '2019-02-26'
  ```

1. Paste them between the marker lines in the code cell below.
1. Fill in your _full name_, _identification number_, and the current _date_ as strings between quotes.
1. Run the code cell by putting the cursor there and typing **Control-Enter**.

In [1]:
#// BEGIN_TODO [Author] Name, Id.nr., Date, as strings (1 point)

AUTHOR_NAME = 'Erin Bolintis'
AUTHOR_ID_NR = '2759817'
AUTHOR_DATE = '2022-9-28'

#// END_TODO [Author]

AUTHOR_NAME, AUTHOR_ID_NR, AUTHOR_DATE

('Erin Bolintis', 'ebo217', '2022-9-28')

## Table of Contents
<div class="toc" style="margin-top: 1em;">
    <ul class="toc-item">
        <li>
            <span><a href="#Introduction-to-this-template-notebook" data-toc-modified-id="Introduction-to-this-template-notebook">Introduction to this template notebook</a></span>
        </li>
        <li>
            <span><a href="#Important-Reminder" data-toc-modified-id="Important-Reminder">Important Reminder</a></span>
        </li>
        <li>
            <span><a href="#1.-Hip-Hip,-Hooray" data-toc-modified-id="1.-Hip-Hip,-Hooray">1. Hip Hip, Hooray</a></span>
        </li>
        <li>
            <span><a href="#2.-Collatz-Problem" data-toc-modified-id="2.-Collatz-Problem">2. Collatz Problem</a></span>
        </li>
        <li>
            <span><a href="#3.-gnitirW-sdrawkcaB" data-toc-modified-id="3.-gnitirW-sdrawkcaB">3. gnitirW sdrawkcaB</a></span>
        </li>
        <li>
            <span><a href="#4.-Writing-Books" data-toc-modified-id="4.-Writing-Books">4. Writing Books</a></span>
        </li>
        <li>
            <span><a href="#5.-Insurance-Calculations" data-toc-modified-id="5.-Insurance-Calculations">5. Insurance Calculations</a></span>
        </li>
        <li>
            <span><a href="#6.-Sorting" data-toc-modified-id="6.-Sorting">6. Sorting</a></span>
        </li>
        <li>
            <span><a href="#7.-Multi-Multiplication<" data-toc-modified-id="7.-Multi-Multiplication<">7. Multi Multiplication</a></span>
        </li>
        <li>
            <span><a href="#8.-String-Cleaning" data-toc-modified-id="8.-String-Cleaning">8. String Cleaning</a></span>
        </li>
        <li>
            <span><a href="#9.-Counting-Letters" data-toc-modified-id="9.-Counting-Letters">9. Counting Letters</a></span>
        </li>
    </ul>
</div>

## Preliminaries

Run the cell below. This cell will import additional modules providing additional Python functionality.

In [2]:
# This set of exercises does not need additional modules

## Important Reminder!

* Follow all coding conventions defined on the Python Coding Standard document available via Canvas. Remember that you are not just programming for a **machine**, you are mainly programming for **other humans!**

* In particular, all function definitions must have **type hints** and a **docstring**. For instance, if you define a function `ABC` that takes the input argument `variable_1` which is an integer and returns a string you need to define the function as follows: 
    
    `def ABC(variable_1: int) -> str:`
  
* If the function `ABC` **would not return anything and only print something (therefore being a void function), the return type would be None**! Therefore:

     `def ABC(variable_1: int) -> None:`

   
* A **valid docstring** starts with a capital letter and ends with a period (.).
* **You need to make sure that the notebook runs through without an error before submitting on CodeGrade**. Otherwise, CodeGrade will not be able to score your notebook correctly!
* The functions that you need to define below to solve the problem should be general, i.e. solving not only the particular example that we provide you for each problem, but should also work with any input value that the input variables can possible take.


## 1. Hip Hip, Hooray

It's your professors birthday!<br>
To celebrate create a function called `hip_hip_hooray` that uses the *integer* variable `new_age` and prints the message, **Hip Hip, Hooray**, as many times as his `new_age` using a a `while-`loop.

**Example:**  
*Input:*  
```python
new_age = 3
```

*Output:*  
<pre>
Hip Hip, Hooray
Hip Hip, Hooray
Hip Hip, Hooray
</pre>

**Important:** Do not forget to use type hints in the function definition!

In [3]:
#// BEGIN_TODO [birthday] Hip Hip, Hooray 

new_age: int=3
def hip_hip_hooray (new_age:int)->None:
    '''Based on the value of new_age it produces hip,hip,hoorays as many times as necesaary.'''
 
    while new_age != 0:
        print("Hip Hip, Hooray")
        new_age -=1
    
hip_hip_hooray(new_age)

        
#// END_TODO [birthday]

Hip Hip, Hooray
Hip Hip, Hooray
Hip Hip, Hooray


## 2. Collatz Problem

We can use `while`-loops to test mathematical problems such as the *collatz* problem.<br>
The *collatz* problem is an easy iterated formula in which you have an *integer* variable called `integer` and has the following formula:
* If the `integer` is even: Divide the integer by 2
* If the `integer` is odd: Multiply the integer by 3 and add 1
* Do this until the sequence reaches 1 (After which it will loop 1, 4, 2, 1, ...)

For this exercise, create a function called `collatz` that uses an *integer* input argument called `integer` and prints out the sequence of the collatz problem (see the example output below).

**Note:** The question of the *collatz* problem is whether every starting value will eventually reach 1. 
This has not been formally proven yet, but it is assumed to be true.

**Hint:** You need to create a program using a `while`-loop that prints the *collatz* sequence of an *integer*.



**Example:**  
*Input:*  
```python
integer = 20
```

*Output:*  
<pre>
10
5
16
8
4
2
1
</pre>

**Important:** Do not forget to use type hints in the function definition!

In [4]:
#// BEGIN_TODO [Collatz] Collatz Problem 
integer: int=20
def collatz(integer:int)-> None:
    '''Applying a while loop to try and solve the collatz problem'''
    while integer != 1:
        if integer % 2 == 0:
            integer = integer // 2
            print(integer)
        else: 
            integer = integer * 3 + 1
            print (integer)
        
            
collatz(integer)

    

#// END_TODO [Collatz]

10
5
16
8
4
2
1


## 3. gnitirW sdrawkcaB

We can use `while`-loops in a function to reverse a string.

For this exercise create a function called `print_backwards` that uses the *string* variable `word` and prints the letters backward.

**Hints:** 
- This function will only print, it does not return anything.
- Use `len(word)` to determine the string's length.
- Use `word[i]` to take the letter at index i.
- Keep in mind that the first character is at index 0.
- Use `print(..., end='')` to suppress printing a new line.

**Example:**  
*Input:*  
```python
word = 'Pizza!'
```

*Output:*  
<pre>
!azziP
</pre>

**Important:** Do not forget to use type hints in the function definition!

In [5]:
#// BEGIN_TODO [reversed] gnitirW sdrawkcaB=
word: str= 'Pizza!'
def print_backwards(word: str) -> None:
    '''Prints the string backwards letter by letter'''
    
    e: int=1
    while e<=len(word):
        print(word[len(word)-e], end='')
        e +=1
    
print_backwards(word)





#// END_TODO [reversed]

!azziP

## 4. Writing Books

We would like to print all chapters and sections for a book.<br>

For this exercise create a function called `print_chapters`, which has `start_chapter`, `final_chapter`, and `max_section` as *integer* parameters:

* `start_chapter` is the first chapter you are going to print.
* `final_chapter` is the final chapter you are going to print.
* `max_section` is the number of sections each chapter has - All chapters have the same number of `max-section`.

**Hint:** `print_chapters` has to print all chapters and sections within each chapter.

**Example:**  
*Input:*  
```python
start_chapter = 1
final_chapter = 3
max_section = 2
```

*Output:*  
<pre>
Chapter 1
Section 1.1 
Section 1.2 
Chapter 2 
Section 2.1 
Section 2.2 
Chapter 3 
Section 3.1 
Section 3.2 
</pre>

**Important:** The function `print_chapters` should take the three input arguments `start_chapter`, `final_chapter` and `max_section` in this order! Do not forget to use type hints in the function definition!

In [6]:
#// BEGIN_TODO [books] Writing Books

start_chapter: int = 1
final_chapter: int = 3
max_section: int = 2

def print_chapters (start_chapter: int, final_chapter:int, max_section: int)-> None:
    '''Prints all chapters and sections for a book'''
    while start_chapter <= final_chapter:
        d: int=1
        print (f"Chapter {start_chapter}")
        while d!=-1:
            print (f"Section {start_chapter}.{max_section-d}")
            d-=1
        start_chapter +=1

print_chapters (start_chapter, final_chapter, max_section)       



#// END_TODO [books]

Chapter 1
Section 1.1
Section 1.2
Chapter 2
Section 2.1
Section 2.2
Chapter 3
Section 3.1
Section 3.2


## 5. Insurance Calculations

We would like to check what price someones car insurance will be.<br>
This is based on the total amount of light and heavy accidents that the person made over the last year, and the number of years of driving experience.

For this exercise, you need to create a function called `checkPrice` that checks the price as often as we want, because we need to check it for multiple people. Your job is to create the function `checkPrice` such that:

* The base insurance is 2,000 a year
* Every light accident increases the price by 100
* Every heavy accident increases the **total price** by 10%

For instance: If the insurance premium was 100 (before any heavy accidents), and one has the first heavy accident, the premium increases by $10\%, i.e. 100 * 1.10 = 110$. The second heavy accident increases the premium then by another $10\%$. Thus, $110 * 1.10 = 121$.

* Every year of driving experience drops the price by 200, but only after the first 5 learner years. Hence, if a driver has 6 years of experience, then her/his insurance payment will be lower by 200 Euros.
* The total price of the insurance can never be lower than 800

**Hint:** `checkPrice` has to return the price of the insurance (What data type has the returned value?)

**Hint:** The function `checkPrice` takes three input arguments, `light_accidents`, `heavy_accidents` and `driving_experience` which are all integer variables!

**Example:**  
*Input:*  
```python
light_accidents = 0
heavy_accidents = 0
driving_experience = 0
```

*Output:*  
<pre>
2000.0
</pre>

**Important:** The function `checkPrice` should take the three input arguments `light_accidents`, `heavy_accidents` and `driving_experience` in this order! Do not forget to use type hints in the function definition!

In [7]:
#// BEGIN_TODO [Insurance] Insurance Calculations
light_accidents: int= 0
heavy_accidents: int= 0
driving_experience: int= 0
    
def checkPrice (light_accidents:int, heavy_accidents:int, driving_experience:int)-> float:
    '''Checks what price someones car insurance will be'''
    base_insurance: float = 2000
    base_insurance = base_insurance+100*light_accidents
    
    if(heavy_accidents!= 0):
        base_insurance = base_insurance+ 1.1**heavy_accidents 
    
    if driving_experience> 5 :
        base_insurance=base_insurance-200*(driving_experience-5)
        
    if base_insurance < 800:
        base_insurance = 800
        
    return float(base_insurance)

checkPrice (light_accidents, heavy_accidents, driving_experience)



#// END_TODO [Insurance]

2000.0

## 6. Sorting

For this exercise create a function called `sorting`, which takes three *integers* variables (see below) as arguments.<br>
This function should sort in ascending order and return the numbers in a string. Numbers should be separated by *commas*, and one *space* is required after each *comma*.

**Note:** 
- This function returns something (a string) and does not print.

**Hint:**
- You should use `max` and `min` functions to solve this problem! 
- You are not allowed to use `if`-statements to solve this problem!

**Example:**  
*Input:*  
```python
num_1 = 4
num_2 = 7
num_3 = 2
```

*Output:*  
<pre>
2, 4, 7
</pre>

**Important:** The function `sorting` should take the input arguments `num_1`, `num_2` and `num_3` in the exact order as outlined above (that is first `num_1` etc.)

In [6]:
#// BEGIN_TODO [sorting_numbers] Sorting 
num_1: int= 4
num_2: int= 7
num_3: int= 2
def sorting (num_1:int, num_2:int, num_3:int)->str:
    '''Sorts in ascending order and returns the numbers in a string'''
    min_1 = min (num_1, num_2, num_3)
    min_2 = min(num_1, num_2)
    min_3 = min(num_2, num_3)
    max_2 = max(min_1, min_2, min_3)
    max_1 = max(num_1, num_2, num_3)
    return f"'{min_1}, {max_2}, {max_1}'"
    
sorting (num_1, num_2, num_3)    


#// END_TODO [sorting_numbers]

"'2, 4, 7'"

## 7. Multi Multiplication

Functions can have all possible variable types as parameters.<br>
This includes *integers*, *floats*, and *strings*. <br>
But functions can also take other functions as parameters.<br>
In order to test this, we are going to create two functions for this exercise:

* The first function called is called `multiply`. This function takes two arguments called `num_1` and `num_2`. It  calculates the product of these two arguments returns the result.

* The second function called is `run_triple`. It will take the function  `multiply` as an argument and besides this, four other arguments, `num_1`, `num_2`, `num_3` and `num_4`. `run_triple` calls the functions `multiply` twice, once with the values `num_1` and `num_2` and once with the arguments `num_3` and `num_4`. Lastly, it will call `multiply` a last time, this time with the two resulting products that one obtained doing the previous two multiplications. Lastly, it will return this final result.

**Hint:** The type to use for functions as arguments is *callable*.

**Note:** For this exercise, we assume that all input arguments will be integers. Hence, the functions will also return integer values!


**Example:**  
*Input:*  
```python
multiply
num_1 = 2
num_2 = 3
num_3 = 3
num_4 = 4
```

*Output:*  
<pre>
72
</pre>

**Important:**: The function should take the arguments in the exact order as outlined above, that is, it should take `multiply` first, then as a second argument `num_1` etc. Do not forget to use type hints in the function definition.

In [8]:
#// BEGIN_TODO [run_triple] Multi Multiplication 

num_1: int= 2
num_2: int= 3
num_3: int= 3
num_4: int= 4
def multiply(num_1:int,num_2:int)->int:
    "Multiplies the two arguments with each other"
    return num_1*num_2

def run_triple(multiply: callable,num_1:int, num_2:int, num_3:int, num_4:int)->int:
    '''Multiplies all arguments with each other, while using the pre-defined multiply function'''
    mr_1 = multiply(num_1,num_2)
    ms_1 = multiply(num_3,num_4)
    return multiply(mr_1,ms_1)


run_triple(multiply, num_1, num_2, num_3, num_4)


#// END_TODO [run_triple]

72

## 8. String Cleaning

Your database is storing the names of customers.<br>
However, because of the way we use these names, the stored names have to follow certain rules. 

For this exercise create a function called `standardize_name` that changes the name if it does not comply to the rules, into names that do. 
Return the changed name, or the original name when it does follow the naming rules.

The rules for the names are:
* Start with capital letter
* All other letters are lowercase
* Only the characters a-z A-Z are allowed in the name, any other letters have to be deleted
* Spaces and tabs are not allowed, underscores are used instead
* Names can be no longer than 80 characters, longer names will lose their extra letters

**Example:**  
*Input:*  
```python
name = "python3 cOUrse"
```

*Output:*  
<pre>
Python_course
</pre>

**Important:** Do not forget to use type hints in the function definition.

In [4]:
#// BEGIN_TODO [transforming_names] String Cleaning 

name: str= "python3 cOUrse"
def standardize_name(name:str)->str:
    '''Cleans the string, so that it complys with the specified rules'''
    new_name=""
    g: int=0
    new: str= name.lower().capitalize()
        
    
    while(g<len(new)):
        if new[g]==" ":
            new_name+="_"
        elif new[g].isalpha():
            new_name += new[g]
        g+=1
  
    return new_name
standardize_name(name)    






#// END_TODO [transforming_names]

'Python_course'

## 9. Counting Letters

For this exercise create a function called `count_case` that accepts a variable called `string` which is of type *string* and returns the number of upper and lower case letters as a string. 

**Example:**  
*Input:*  
```python
"I am currently attending Programing Course"
```

*Output:*  
<pre>
Upper case: 3 - Lower case: 34
</pre>

**Important:** Do not forget to use type hints in the function definition. Also, be aware of whitespaces and capital and small letters in the final string output!

In [11]:
#// BEGIN_TODO [count_case] Counting Letters 

string: str= "I am currently attending Programing Course"
def count_case(string:str)->str:
    '''Counts all upper and lower case letters and returns it as a string'''
    i: int=0
    up: int=0
    down: int=0
        
    while i<len(string):
        if string[i].isupper():
            up+=1
            
        elif string[i].islower():
            down+=1
            
        i+=1
        
    return f"Upper case: {up} - Lower case: {down}"

count_case(string)




#// END_TODO [count_case]

'Upper case: 3 - Lower case: 34'

---

# (End of Notebook)

&copy; 2020-2022 - **TU/e** - **VU Amsterdam**